### ISSUE TITLE
Incompletenesses of guard condition in the presence of cofix redexes or not enough immediately applied primitive projection

### ISSUE THREAD

**Issue by herbelin** [coq/coq#15451]
#### Description of the problem

Here are two examples of fixpoints whose normal form is guarded but an unreduced cofix redex in the way blocks the normalization.

```coq
CoInductive Stream := {hd : nat; tl : Stream}.
Definition zeros := cofix zeros := {|hd := 0; tl := zeros|}.

Fixpoint f n :=
  match n with
  | 0 => 0
  | S n =>
    match zeros with
    | {|hd:=_|} => fun f => f n
    end f
  end.
```

```coq
Set Primitive Projections.
CoInductive Stream A := {hd : A; tl : Stream A}.
Arguments hd {A} s.

Fixpoint f n :=
  match n with
  | 0 => 0
  | S n => (cofix cst := {|hd := (fun f => f n); tl := cst|}).(hd) f
  end.
```

Found while working on #15434.

#### Coq Version

At least since Coq 8.3 (and probably 8.0) for the former, 8.5 for the latter, with #9602 in 8.10 having fixed most of preexisting incompletenesses.

**Comment by herbelin** [comment]
Here is another issue found while working on #15434. The stack is discarded when reducing primitive projection in guard condition:

```coq
Set Primitive Projections.

Record T := { a : nat -> nat }.

(* Works ok *)
Check fix f (b:bool) n : nat :=
  match n with
  | 0 => 0
  | S n => (if b then f b else f b) n
  end.

(* Fail but should not *)
Check fix f (b:bool) n : nat :=
  match n with
  | 0 => 0
  | S n => (if b then {| a := f b |}.(a) else f b) n
  end.
```

I don't see how it could be critical though, since missing arguments on an empty stack are considered "not subterm".

**Comment by herbelin** [comment]
Fixed in #15498.
